### PR TITLE
Add support for cargo bench (Fixes #9)

### LINKED ISSUE #9: Does not work with cargo bench
After installing and trying `cargo with "echo {bin}" -- bench`
```
Error: ErrorMessage { msg: "Unable to convert \'bench\' into a cargo subcommand" }

stack backtrace:
   0: failure::backtrace::internal::InternalBacktrace::new::h59b3db8423e4cf5e (0x10e71c9b8)
   1: <failure::backtrace::Backtrace as core::default::Default>::default::ha057354baab460c1 (0x10e71cc64)
   2: cargo_with::cargo::Cmd::from_strs::h41a41d93068aa265 (0x10e7123cd)
   3: cargo_with::runner::runner::h426b4b78cc407ce9 (0x10e717517)
   4: cargo_with::main::h57a9494064ed692b (0x10e70d932)
   5: std::rt::lang_start::{{closure}}::h6a72812a9432124a (0x10e71663a)
   6: std::panicking::try::do_call::hb3270f6ae846849d (0x10e80f837)
   7: ___rust_maybe_catch_panic (0x10e8215be)
   8: std::rt::lang_start_internal::ha07e743882e341f3 (0x10e8101ca)
   9: _main (0x10e70daf8)
```

### PR DESCRIPTION
**Pull request by cbourjau**
This should fix the issue at hand but I might have killed some corner
cases. I deleted some error messages which don't look as if they were
actually sound anyways. What do you think, @barskern ?

### CODE REVIEW

File: src/cargo.rs [hunk 1]
```diff
@@ -194,48 +197,34 @@ pub(crate) fn select_buildopt<'a>(
...
-        } else {
-            Err(format_err!("Found more than one possible candidate:\n\n{}\n\nPlease use `--example` or `--bin` to specify exactly what binary you want to examine", candidates_str))?
-        }
+        .collect();
+    // We expect exactly one candidate; everything else is an error
+    match candidates.as_slice() {
+        [] => Err(err_msg("No suitable build artifacts found.")),
+        [the_one] => Ok(the_one),
+        the_many => Err(format_err!(
+            "Found several artifact candidates: \n :{:?}",
```
Comment (barskern) [inline src/cargo.rs:224]: I think we should still print an error messages which suggests ways to solve the error (when many artifacts are found). E.g "Please use `--test` [...]".

File: src/cargo.rs [hunk 2]
```diff
@@ -194,48 +197,34 @@ pub(crate) fn select_buildopt<'a>(
...
-            .collect();
-
-        if is_test {
-            Err(format_err!("Found more than one possible candidate:\n\n{}\n\nPlease use `--test`, `--example`, `--bin` or `--lib` to specify exactly what binary you want to examine", candidates_str))?
-        } else {
-            Err(format_err!("Found more than one possible candidate:\n\n{}\n\nPlease use `--example` or `--bin` to specify exactly what binary you want to examine", candidates_str))?
-        }
+        .collect();
+    // We expect exactly one candidate; everything else is an error
+    match candidates.as_slice() {
```
Comment (barskern) [inline src/cargo.rs:220]: Hmm I haven't tested it but do we really need the `as_slice` call here? No matter what it is a bit easier on the eyes than `&*candidates` anyways :smile:

File: src/cargo.rs [hunk 3]
```diff
@@ -194,48 +197,34 @@ pub(crate) fn select_buildopt<'a>(
...
-            // When run as a test we only care about the binary where the profile
-            // is set as `test`
-            if is_test {
-                opt.profile.test
-            } else {
-                opt.target
+            // When run as a test or bench we only care about the
+            // binary where the profile is set as `test`
+            match cmd_kind {
+                CmdKind::Test | CmdKind::Bench => opt.profile.test,
```
Comment (barskern) [inline src/cargo.rs:210]: I'm guessing you tested running it and the test property was set to true when running benchmarks?

**Review by barskern — COMMENTED** [review]
This is a nice refactor and feature!

I think the new error message is a bit generic, what do you think?

File: src/cargo.rs (hunk 2 above)
Comment (cbourjau) [reply src/cargo.rs:220]: Its needed in this case at least

File: src/cargo.rs (hunk 3 above)
Comment (cbourjau) [reply src/cargo.rs:210]: Indeed, turns out having small projects is good for testing (see #11 ). Who would have thought :D

**Comment by barskern** [comment]
We should probably also add some info about this in the README

File: src/cargo.rs (hunk 1 above)
Comment (cbourjau) [reply src/cargo.rs:224]: I am out of time for today, but will look into this in the next few days

**Comment by cbourjau** [comment]
Thanks already, I'm out of time for today, but will take a closer look at those `--test` flags in the coming days.

**Comment by velvia** [comment]
@cbourjau thanks for doing this!   This will be very helpful to many of us hoping to integrate cargo bench with other profiling tools.

**Comment by barskern** [comment]
@cbourjau if you want we could merge this now and I could fix up the error message in a separate PR? I fully understand that you have other things which take time. As I currently have some available time, I wouldn't mind doing it. 😊

**Comment by cbourjau** [comment]
Yes, please go ahead! Thanks a lot!

Ole Martin Ruud writes:

> @cbourjau if you want we could merge this now and I could fix up the error message in a separate PR? I fully understand that you have other
> things which take time. As I currently have some available time, I wouldn't mind doing it. 😊
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.


-- 
Christian Bourjau
+45 915 666 01 | c.bourjau@posteo.de
LinkedIn: http://se.linkedin.com/pub/christian-bourjau/93/888/77/

**Pull request merged**